tiledb/sm/query/result_tile.cc

```diff
@@ -910,14 +909,248 @@ void ResultTile::compute_results_count_sparse(
+      // in this partition.
+      const char* const first_c_coord = &buff_str[first_c_offset];
+      const char* const second_coord = &buff_str[last_c_offset];
+
+      // Check if the first and last coordinates have an identical
+      // string value.
+      if (first_c_size == last_c_size &&
+          strncmp(first_c_coord, second_coord, first_c_size) == 0) {
+        const uint8_t intersects = str_coord_intersects(
+            first_c_offset, first_c_size, buff_str, range_start, range_end);
+        if (intersects) {
+          for (uint64_t pos = first_c_pos; pos < first_c_pos + c_partition_size;
+               ++pos) {
+            r_bitmap[pos] = 1;
+          }
+        }
+      } else {
+        // Compute results
+        uint64_t c_offset = 0, c_size = 0;
+        for (uint64_t pos = first_c_pos; pos <= last_c_pos; ++pos) {
+          c_offset = buff_off[pos];
+          c_size = (pos < coords_num - 1) ? buff_off[pos + 1] - c_offset :
+                                            buff_str_size - c_offset;
+          r_bitmap[pos] = str_coord_intersects(
+              c_offset, c_size, buff_str, range_start, range_end);
+        }
+      }
+    }
+
+    // We've calculated `r_bitmap` for all coordinate values in
+    // the sorted dimension.
+    return;
+  }
+
+  // Here, we're computing on all other dimensions. After the first
+  // dimension, it is possible that coordinates have already been determined
+  // to not intersect with the range in an earlier dimension. For example,
+  // if `has_previous_dim` is true, we have already checked this coordinate
+  // for an intersection in previous dimensions. If either did not intersect
+  // a coordinate, its associated value in `r_bitmap` will be zero. To optimize
+  // for the scenario where there are many contiguous zeroes, we can `memcmp`
+  // a large range of `r_bitmap` values to avoid a comparison on each
+  // individual element.
+  //
+  // Often, a memory comparison for one byte is as quick as comparing
+  // 4 or 8 bytes. We will only get a benefit if we successfully
+  // find a `memcmp` on a much larger range.
+  const uint64_t zeroed_size = coords_num < 256 ? coords_num : 256;
+  for (uint64_t i = 0; i < coords_num; i += zeroed_size) {
+    const uint64_t partition_size =
+        (i < coords_num - zeroed_size) ? zeroed_size : coords_num - i;
+
+    // Check if all `r_bitmap` values are zero between `i` and
+    // `partition_size`.
+    if (has_previous_dim && r_bitmap[i] == 0 &&
+        memcmp(
+            &r_bitmap[i],
+            &r_bitmap[i + 1],
+            (partition_size - 1) * sizeof(uint8_t)) == 0)
+      continue;
+
+    {
+      // Here, we know that there is at least one `1` value within the
+      // `prev_dim_r_count` values within this partition. We must check
+      // each value for an intersection.
+      uint64_t c_offset = 0, c_size = 0;
+      for (uint64_t pos = i; pos < i + partition_size; ++pos) {
```

Comment:
This is telling me that the parallelization happens only across tiles, not dimensions. Another idea is to create one bitmap per tile per dimensions, run the filtering in parallel, and then OR the bitmaps together in a separate subsequent function call.

Reply:
We could do this, but I'm wondering about the memory overhead?

Reply:
https://app.shortcut.com/tiledb-inc/story/12133/sparse-global-order-reader-advanced-optimizations-decide-to-parallelize-on-tiles-dimensions-or-ranges